### ISSUE TITLE
Handling multiple runs

### ISSUE THREAD

**Issue by adelavega** [tyarkoni/coda#6]
An existing difference between BIDSEventReader and EventReader (FSL-style), is how runs are handled. In the latter, there seems to be some functionality for trying to guess run number from the path, and it is coded as a columns. Currently BIDSEventReader just handles one file at the time, so it assumes you're dealing w/ meta-data outside of coda.

However, when wrapping coda with a Nipype interface, it's not clear how this should be handled. I *think* in your pull request as it current works, SpecifyEvents just lumps all event files for a subject (that match the specified runs), and I don't think run is coded in the output. This probably needs to be added, since SpecifyModel expects a list of Bunches, where each Bunch specifies the model for each run. 

I suppose if multiple bids events file are given, it's easy enough to just loop over them and return a list of bunches, rather than one. However for FSL-style event files, we may have to add the functionality to match run pattern to be able to tell apart which events came from which run, unless I'm misreading your code.

**Comment by adelavega** [comment]
I'm starting to implement this for BIDS files, with the assumption that the Transformer should make transformations on all runs at once (e.g. you want to standardize RT w.r.t all runs, not just one). There may be some issues here, as condition may only appear in some runs (e.g. incorrect trials). Not clear if nipype expects every run to declare all possible conditions, or only those in the run. Im guessing the latter and need to find out.

**Comment by tyarkoni** [comment]
Coda doesn't have any concept of a run or a subject; it operates on whatever is handed to it. If it gets multiple event files, it concatenates them and treats them as one. So in that sense, whether it applies transformations to all runs at once or one at a time just depends on how they're passed in.

From a Nipype perspective, things will be more complicated, as the SpecifyEventsInterface will sometimes need to be invoked strictly a run-by-run basis (which is easy), and other times it will need to be handed multiple runs at once, or even multiple subjects at once, in order to estimate parameters to apply to individual runs. But let's deal with that in the Nipype interface that wraps coda, not in coda itself.

**Comment by adelavega** [comment]
Right. Well I made some commits that sort of get at this (most recent ones). The issue is the way EventTransformer represents internally the data, it mashes it all together. So if one were to feed in multiple runs, when it resamples to TR it would average the two runs. 

What I did is add a "run" column, which is basically just an identifier for a grouping of events (i.e. different event.tsv files). Then, I added a groupby operation on resample so it resamples each run separately. Finally, I added a method to get the data, and it can either pass it all back as is, or split it up by "run". Maybe we can change this terminology to just be "group". That way, you could pass multiple event files (in a list of lists), and be able to apply operations to all, but still keep them separate. 

I'm open to reworking this completely, just wanted to get something would allow me to have a complete first level work flow for now (which I do now).

**Comment by adelavega** [comment]
Oh I'll also point out I forked nipype and added "issues" to my repo (separate from nipype's issues), so I'm keep track of some notes/to-do there, FYI

**Comment by tyarkoni** [comment]
The way it would probably work is that we'd write new transformer methods that just return the summary statistics we need, and then those would be passed in as arguments when calling operations like `standardize`. I think the alternative of making coda run- and subject-aware would get very messy, because in principle we want to be able to extend the same approach to arbitrary levels of analysis (e.g., maybe someone wants to standardize RTs based on entire sessions containing multiple subjects).

To make it a bit more concrete, say we have trial-by-trial emotion ratings for each subject for each stimulus, and we want to add a new column to the design matrix that codes the normative rating for each stimulus across all subjects. Then, inside the Nipype interface, we would feed in all subjects' event files to coda in one shot, and call some kind of `aggregate` transformation that allows us to specify that we want to group on stimulus id (we could even do the aggregation operation in the interface code rather than coda itself, and only use coda for reading and concatenating all the events into a DF). This would return a dict of stimulus id => mean rating mappings. Then, we would loop over each subject (or run) as per normal, except we would now pass in the extracted mean values for the stimuli, and call some kind of `inject` transformation (currently not implemented) that would map the incoming values to rows in the run-specific event file based on the index key (in this case, stimulus ID).

The bottom line is that I don't think we should add logic to coda to make it internally aware of things like runs, subjects, etc. That should all be handled inside the Nipype interface and/or by adding cleverly constructed transformations to coda (i.e., that take the right arguments).

**Comment by adelavega** [comment]
Gotcha. I can see how it would get out of hand pretty quickly, especially with subjects, hadn't thought of that. I have some remaining questions/comments that I'll note here but might be better to discuss in person. 

1) So summary statistics like n, mean and std, etc, that would allow us to standardize relative to the mean / std across runs, for example? Would you expect these summary statistics operations to be described in the transformation.json file so they can be applied using the Transformer's "apply_from_json" method? Or alternatively we can do this in the nipype interface. 

2) Minor: given what you said, I'm assuming then if a BIDSEventReader is given multiple event files, it should concatenate them into one dataframe. Right now it returns a list of dataframes.

**Comment by adelavega** [comment]
Add aggregation functionality to coda (e.g. mean, max, min, std)

**Comment by adelavega** [comment]
Update: applying within transformation is going just perfectly fine, but have no implemented across-run transformations, summarization or injection.
